src/node/transaction.cpp

```diff
@@ -71,6 +71,12 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
 
         // Transaction was accepted to the mempool.
 
+        if (relay) {
+            // the mempool tracks locally submitted transactions to make a
+            // best-effort of initial broadcast
```

Comment:
Previously, we would also best-effort rebroadcast (meaning not initial). After this change, best-effort applies only to the initial broadcast indeed (or the absent-from-mempool broadcast).

Is this the expected rebroadcast behavior? Do we clearly communicate that software-triggered-rebroadcast (wallet or whatever) won't be assisted by the node if the node was offline at the time in the same way the node assists initial broadcast.

Behavior change:
1. A transaction added to the mempool initially broadcasted
2. Node goes offline
3. A transaction is still in the mempool, triggers rebroadcast.
4. Rebroadcast fails because offline

The difference is that after this code, UnbroadcastTx won't help in step 4.
Indeed, the peers are unlikely to request it again. It's still possible though, if our mempool is more persistent (e.g., deeper in size).

This applies backward too: if our mempool is less persistent, we will make useless rebroadcasts (peers won't send GETDATA because they still have it in the mempool). This part is indeed not a degradation over master, but still something to think about. Small-mempool  nodes may become are likely to become "10-15 minutes" rebroadcast spammers (after only one extra broadcast trigger, but because Unbroadcast set never clears)

Reply:
> Previously, we would also best-effort rebroadcast (meaning not initial). After this change, best-effort applies only to the initial broadcast indeed (or the absent-from-mempool broadcast).
>
> Is this the expected rebroadcast behavior?

Yes, this is the intention, and was the original intention behind the unbroadcast set. Locally submitted transactions should be tracked to make sure that at least one peer had requested them _at least once_.

> Behavior change:
>
> 1. A transaction added to the mempool initially broadcasted
> 2. Node goes offline
> 3. A transaction is still in the mempool, triggers rebroadcast.
> 4. Rebroadcast fails because offline
>
> The difference is that after this code, UnbroadcastTx won't help in step 4.

Correct. That's intentional. The eventual aim of the rebroadcast project is to remove the need for the local wallet to rebroadcast transactions in the majority of cases. I hope that eventually the default behavior is that (3) doesn't happen at all.